### DISCUSSION TITLE
Does it support aarch64 architecture？

### DISCUSSION THREAD

**Discussion by lifan95** [micropython/micropython#9377] [Using MicroPython]
COMMANDS:
cd mpy-cross
make clean
make
cd ../ports/unix
make clean
make CROSS_COMPILE=aarch64-linux-gnu- deplibs
make CROSS_COMPILE=aarch64-linux-gnu- MICROPY_STANDALONE=1 MICROPY_PY_BTREE=0 MICROPY_PY_USSL=1 CFLAGS_EXTRA=-DMICROPY_GCREGS_SETJMP=1 CWARN= 

ERROR MESSAGES:
CC ../../py/nlraarch64.c
/tmp/ccuhWd3l.s: Assembler messages:
/tmp/ccuhWd3l.s:10: Error: operand 1 must be a floating-point register -- `stp lr,x9,[x0,#16]'
/tmp/ccuhWd3l.s:65: Error: operand 1 must be a floating-point register -- `ldp lr,x9,[x1,#16]'
../../py/mkrules.mk:77: recipe for target 'build-standard/py/nlraarch64.o' failed
make: *** [build-standard/py/nlraarch64.o] Error 1

**Comment by jimmo** [comment]
I just tested and it's working here fine.

I ran `make CROSS_COMPILE=aarch64-linux-gnu- MICROPY_PY_FFI=0` and it compiled successfully.

Then

```
$ qemu-aarch64 -L /usr/aarch64-linux-gnu ./build-standard/micropython
MicroPython v1.19.1-438-gffafb1f34-dirty on 2022-09-21; linux [GCC 12.2.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> 1+1
2
```

Which version of aarch64-linux-gnu-gcc do you have? Mine is 12.2.

**Reply by lifan95** [reply]
Very pleased to receive your reply.

Mine is 7.4.1.

**Reply by jimmo** [reply]
Can you try upgrading?

**Reply by lifan95** [reply]
Thanks!

I upgraded the version of aarch64-linux-gnu-gcc to 12.2.1 and it compiled successfully. But I'm not sure from which version it can be successfully compiled.

In addtion, there seems to be a small problem when MICROPY_PY_FFI=1 and MICROPY_STANDALONE=1 specified. Compilation will report the following error:
`aarch64-linux-gnu/bin/ld: cannot find build-standard/lib/libffi/out/lib/libffi.a: No such file or directory` 

Compilation will be ok after I modify `ports/unix/Makefile` as follows:
```
ifeq ($(MICROPY_PY_FFI),1)

ifeq ($(MICROPY_STANDALONE),1)
LIBFFI_CFLAGS_MOD := -I$(shell ls -1d $(BUILD)/lib/libffi/out/lib/libffi-*/include)
 ifeq ($(MICROPY_FORCE_32BIT),1)
  LIBFFI_LDFLAGS_MOD = $(BUILD)/lib/libffi/out/lib32/libffi.a
 else
#  LIBFFI_LDFLAGS_MOD = $(BUILD)/lib/libffi/out/lib/libffi.a
  LIBFFI_LDFLAGS_MOD = $(BUILD)/lib/libffi/out/lib64/libffi.a
 endif
else
LIBFFI_CFLAGS_MOD := $(shell pkg-config --cflags libffi)
LIBFFI_LDFLAGS_MOD := $(shell pkg-config --libs libffi)
endif
```
 
Please help to confirm whether this is a problem.

**Reply by jimmo** [reply]
> Please help to confirm whether this is a problem. 
@lifan95 see 
https://github.com/libffi/libffi/issues/769

**Comment by lifan95** [comment]
aarch64-linux-gnu-gcc7.4.1 doesn't seem to support the register alias lr for x30, so use the x names.
![use_x30_instead_of_register_alias_lr_for_lower_versions_of_aarch64-linux-gnu-gcc](https://user-images.githubusercontent.com/103096326/191493910-181aa18f-915a-41b5-be9c-d5a5559f198f.png)
